### PR TITLE
SASS-2467 - Create Pension Scheme Tax Reference (PSTR) page

### PR DESCRIPTION
**Pull request by srinivaspach**
### Description
Please include a brief summary of the change you've made and the issue that it fixes.

[SASS-2467](https://jira.tools.tax.service.gov.uk/browse/SASS-2467)

##### Before Reviewing
- [ ]  Have you pulled the branch down?
- [ ]  Have you assigned yourself to the PR?
- [ ]  Have you moved the task to “in review” on JIRA?
- [ ]  Have you checked to ensure all dependencies are up to date?
- [ ]  Have you checked to ensure its been rebased against the current version of main?

##### Whilst Reviewing
- [ ]  Have you run the tests?
- [ ]  Have you run the journey tests?
- [ ]  Have you looked at the JIRA story to make sure all Acceptance Criteria has been met?

##### After Reviewing
- [ ]  Have you checked for merge conflicts or any changes in the current main that may affect the current pull request? i.e. does it need another rebase?
- [ ]  Have you checked to make sure there are no builds in the pipeline before you merge?
- [ ]  Have you moved the task to “in pipeline” on Jira?

##### Before creating PR
- [X]  Have you run the tests?
- [ ]  Have you run the journey tests? (where applicable)
- [X]  Have you addressed warnings where appropriate?
- [X]  Have you rebased against the current version of main?
- [ ]  Have you checked code coverage isn’t lower than previously?

##### After PRs been raised
- [ ]  Have you checked the PR Builder passes?

### CODE REVIEW

File: conf/messages [hunk 1]
```diff
@@ -407,12 +411,24 @@ pensions.annualAllowanceTaxPaidByPensionProviderAmount.error.overMaximum.agent =
...
+unauthorisedPayments.pension.pensionSchemeTaxReference.paragraph1.individual = You can get this information from your pension provider.
+unauthorisedPayments.pension.pensionSchemeTaxReference.paragraph1.agent = Your client can get this information from their pension provider.
+unauthorisedPayments.pension.pensionSchemeTaxReference.paragraph2.individual = If you got unauthorised payments from more than one UK pension provider, you can add the references later.
+unauthorisedPayments.pension.pensionSchemeTaxReference.paragraph2.agent = If your client got unauthorised payments from more than UK pension provider, you can add the references later.
+unauthorisedPayments.pension.pensionSchemeTaxReference.whatIsPSTR = What is the PSTR?
+unauthorisedPayments.pension.pensionSchemeTaxReference.error.noEntry.individual = Enter a Pension Scheme Tax Reference
+unauthorisedPayments.pension.pensionSchemeTaxReference.error.incorrectFormat.individual = Enter the Pension Scheme Tax Reference in the correct format
+unauthorisedPayments.pension.pensionSchemeTaxReference.error.noEntry.agent = Enter your client’s Pension Scheme Tax Reference
```
Comment (abiodxn) [inline conf/messages:429]: The google doc for the errors doesn't show a difference between individual and agent. in fact they both use the same error message

File: app/forms/UnauthPaymentsPensionSchemeTaxReferenceForm.scala [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+import filters.InputFilters
+import forms.validation.StringConstraints.validateChar
+import forms.validation.mappings.MappingUtil.trimmedText
+import forms.validation.utils.ConstraintUtil.ConstraintUtil
+import play.api.data.Form
+import play.api.data.validation.Constraint
+import play.api.data.validation.Constraints.nonEmpty
+
+object UnauthPaymentsPensionSchemeTaxReferenceForm extends InputFilters {
```
Comment (abiodxn) [inline app/forms/UnauthPaymentsPensionSchemeTaxReferenceForm.scala:27]: This form isn't needed. You can use the `PensionSchemeTaxReferenceForm` and refactor it slightly so the arguments will accept the error messages.

File: app/controllers/pensions/unauthorisedPayments/UnauthorisedPensionSchemeTaxReferenceController.scala [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                                                                errorHandler: ErrorHandler,
+                                                                clock: Clock,
+                                                                ec: ExecutionContext) extends FrontendController(cc) with I18nSupport {
+
+
+  def show(taxYear: Int): Action[AnyContent] = (authAction andThen taxYearAction(taxYear)).async{ implicit request =>
+    pensionSessionService.getPensionSessionData(taxYear, request.user).flatMap{
+      case Left(_) => Future.successful(errorHandler.handleError(INTERNAL_SERVER_ERROR))
+      case Right(Some(pensionsUserData)) => {
+        val emptyForm: Form[String] = PensionSchemeTaxReferenceForm.pensionSchemeTaxReferenceForm(request.user.isAgent)
```
Comment (abiodxn) [inline app/controllers/pensions/unauthorisedPayments/UnauthorisedPensionSchemeTaxReferenceController.scala:49]: At the moment we don't have the capability to add or reference a particular pension scheme tax reference, however i think you should add a TODO here to remind the person implementing the next story that it needs to be updated here

File: it/controllers/pensions/unauthorisedPayments/UnauthorisedPensionSchemeTaxReferenceControllerISpec.scala [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+          textOnPageCheck(user.specificExpectedResults.get.expectedParagraph1, paragraphSelector(1))
+          textOnPageCheck(user.specificExpectedResults.get.expectedParagraph2, paragraphSelector(2))
+          textOnPageCheck(hintText, hintTextSelector)
+          inputFieldValueCheck(inputName, inputSelector, "")
+          buttonCheck(expectedButtonText, continueButtonSelector)
+          formPostLinkCheck(pensionSchemeTaxReferenceUrl(taxYearEOY), formSelector)
+          welshToggleCheck(user.isWelsh)
+        }
+
+        "render the 'PSTR' page with correct content with pre-filling and a PSTR index" which {
```
Comment (abiodxn) [inline it/controllers/pensions/unauthorisedPayments/UnauthorisedPensionSchemeTaxReferenceControllerISpec.scala:158]: How does this work ? As we don't have the ability to use a pstr index as its not been implemented yet

**Review by abiodxn — CHANGES_REQUESTED** [review]

File: conf/messages (hunk 1 above)
Comment (erasmos-hmrc) [reply conf/messages:429]: We've confirmed that the message for agents [should be](https://docs.google.com/spreadsheets/d/1zC9rsIlxbF1JJb17KxMaa17QMxqZczPdbs0KZkibXFw/edit#gid=1137144805):

> Enter your client's Pension Scheme Tax Reference in the correct format

File: app/controllers/pensions/annualAllowance/PensionSchemeTaxReferenceController.scala [hunk 5]
```diff
@@ -43,8 +43,23 @@ class PensionSchemeTaxReferenceController @Inject()(implicit val cc: MessagesCon
...
+        "pension.pensionSchemeTaxReference.error.noEntry.individual",
+        "pension.pensionSchemeTaxReference.error.incorrectFormat.individual"
+      )
+    }
+  }
+
   private def prefillValue(pstrListOpt: Option[Seq[String]], pensionSchemeTaxReference: Option[Int], user: User): Form[String] = {
-    val emptyForm: Form[String] = PensionSchemeTaxReferenceForm.pensionSchemeTaxReferenceForm(user.isAgent)
+    val errorMsgDetails = formDetails(user.isAgent)
+    val emptyForm: Form[String] = PensionSchemeTaxReferenceForm.pensionSchemeTaxReferenceForm(errorMsgDetails._1, errorMsgDetails._2)
```
Comment (abiodxn) [inline app/controllers/pensions/annualAllowance/PensionSchemeTaxReferenceController.scala:62]: Not needed to be so verbose, you can do something like this 
"(s"pension.pensionSchemeTaxReference.error.noEntry.${if (isAgent) "agent" else "individual"}" and then you could even do that inline and not have to write the function at all.

File: app/controllers/pensions/unauthorisedPayments/UnauthorisedPensionSchemeTaxReferenceController.scala [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                                                                authAction: AuthorisedAction,
+                                                                pensionSchemeTaxReferenceView: PensionSchemeTaxReferenceView,
+                                                                appConfig: AppConfig,
+                                                                pensionSessionService: PensionSessionService,
+                                                                errorHandler: ErrorHandler,
+                                                                clock: Clock,
+                                                                ec: ExecutionContext) extends FrontendController(cc) with I18nSupport {
+
+  def formDetails(isAgent : Boolean) : (String, String) = {
+    if (isAgent) {
```
Comment (abiodxn) [inline app/controllers/pensions/unauthorisedPayments/UnauthorisedPensionSchemeTaxReferenceController.scala:45]: same as above

**Review by abiodxn — CHANGES_REQUESTED** [review]

File: app/controllers/pensions/annualAllowance/PensionSchemeTaxReferenceController.scala (hunk 5 above)
Comment (erasmos-hmrc) [reply app/controllers/pensions/annualAllowance/PensionSchemeTaxReferenceController.scala:62]: It would be enough to have a val and interpolate that.

**Review by abiodxn — APPROVED** [review]

**Pull request merged**